web-wallet/src/lib/components/Allocate/Allocate.svelte

```diff
@@ -54,17 +53,49 @@
 
   let { gasLimit, gasPrice } = gasSettings;
 
-  /** @type {number} */
-  let shieldedAmount = luxToDusk(shieldedBalance);
-
-  /** @type {number} */
-  let unshieldedAmount = luxToDusk(unshieldedBalance);
-
   /** @type {number} */
   let screenWidth = window.innerWidth;
 
   const minAmount = 0.000000001;
 
+  // Internal state of the balances
+  let shielded = shieldedBalance;
+  $: unshielded = totalBalance - shielded;
+
+  // Constant total
+  const totalBalance = shieldedBalance + unshieldedBalance;
+
+  // Used to keep the difference between the initial shielded balance and the current one
+  const initialShielded = shieldedBalance;
+
+  // Derived number states for UI Inputs
+  $: shieldedNumber = luxToDusk(shielded);
+  $: unshieldedNumber = luxToDusk(unshielded);
+
+  let inputsInvalid = false;
+
+  // @ts-ignore
+  function handleShieldedChange(event) {
+    const value = parseFloat(event.target.value);
+    if (isNaN(value) || value < 0) {
+      inputsInvalid = true;
+      return;
+    }
+    inputsInvalid = false;
+    shielded = duskToLux(value);
+  }
+
+  // @ts-ignore
+  function handleUnshieldedChange(event) {
+    const value = parseFloat(event.target.value);
+    if (isNaN(value) || value < 0) {
+      inputsInvalid = true;
+      return;
+    }
+    inputsInvalid = false;
+    shielded = totalBalance - duskToLux(value);
+  }
```

Comment:
Suggestion: these two functions can be easily merged into one.
For example by giving text inputs a name and calculating the `shielded` value based on the event target.